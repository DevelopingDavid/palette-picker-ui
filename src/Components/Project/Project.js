import React, { Component } from 'react';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import Collapse from '@material-ui/core/Collapse';
import List from '@material-ui/core/List';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';

const styles = theme => ({
  root: {
    width: '100%',
    maxWidth: 360,
    backgroundColor: theme.palette.background.paper,
  },
  nested: {
    paddingLeft: theme.spacing.unit * 4,
  },
});

class Project extends Component {
  constructor(props) {
    super(props);
    this.state = {
      palettes: [],
      open: false
    }
  }

  toggleDrawer = () => {
    this.setState(state => ({ open: !state.open }));
  };

  componentDidMount() {
    this.fetchPalette();
  }

  fetchPalette = async () => {
    const { project } = this.props;
    const id = project.id;
    const response = await fetch(`http://localhost:3001/api/v1/projects/${id}/palettes`);
    const data = await response.json();
    this.setState({ palettes: data });
  }

  render() {
    const { classes } = this.props;
    const { project } = this.props;
    return (
      <div>
        <ListItem button onClick={this.toggleDrawer}>
          <ListItemIcon>{<i className="material-icons">palette</i>}</ListItemIcon>
          <ListItemText inset primary={project.project_name} />
          {this.state.open ? <ExpandLess /> : <ExpandMore />}
        </ListItem>
        <Collapse in={this.state.open} timeout="auto" unmountOnExit>
            {this.state.palettes.map((palette) => (
              <List component="div" disablePadding>
                <Divider />
                  <ListItemIcon>{<i className="material-icons">style</i>}</ListItemIcon>
                  <ListItem className={classes.nested}>
                    <ListItemText primary={palette.color_one} style={{backgroundColor: palette.color_one}} />
                  </ListItem>
                  <ListItem className={classes.nested}>
                    <ListItemText primary={palette.color_two} style={{backgroundColor: palette.color_two}} />
                  </ListItem>
                  <ListItem className={classes.nested}>
                    <ListItemText primary={palette.color_three} style={{backgroundColor: palette.color_three}} />
                  </ListItem>
                  <ListItem className={classes.nested}>
                    <ListItemText primary={palette.color_four} style={{backgroundColor: palette.color_four}}/>
                  </ListItem>
                  <ListItem className={classes.nested}>
                    <ListItemText primary={palette.color_five} style={{backgroundColor: palette.color_five}}/>
                  </ListItem>
                <Divider />
              </List>
            ))}
        </Collapse>
      </div>
    )
  }
}

Project.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Project);


