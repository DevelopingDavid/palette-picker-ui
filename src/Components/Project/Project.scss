.palette-icon {
  margin-left: 15px;
  margin-top: 10px;
}

.delete-container {
  margin: 5px 0 15px 30px;
}

#empty-project {
  padding: 10px 0;
}