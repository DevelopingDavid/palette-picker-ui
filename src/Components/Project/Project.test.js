import React from 'react'
import { Project, mapStateToProps, mapDispatchToProps } from './Project';
import { createShallow } from '@material-ui/core/test-utils';
import shortid from 'shortid';

let mockProject = {
  id: 57,
  project_name: "Project 1"
};

let mockFunction = jest.fn();

describe('Project', () => {
  let wrapper;
  let shallow;
  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(
      <Project project={mockProject} fetchProjectsThunk={mockFunction} />
    )
  });

  it('should match Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state on toggleDrawer', () => {
    wrapper.instance().toggleDrawer();

    expect(wrapper.state('open')).toEqual(true);
  });

  it('should call upon fetchPalette on toggleDrawer', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchPalette');
    instance.toggleDrawer();

    expect(instance.fetchPalette).toHaveBeenCalled();
  });

  it('should fetch all palletes & setState', () => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });

    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    wrapper.instance().fetchPalette();
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(`http://localhost:3001/api/v1/projects/${mockProject.id}/palettes`);
  });

  it('should delete project', () => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    const mockOptions = {method: 'DELETE'};
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    wrapper.instance().deleteProject();

    expect(global.fetch).toHaveBeenCalled();
    expect(global.fetch).toHaveBeenCalledWith(`http://localhost:3001/api/v1/projects/${mockProject.id}`, mockOptions);
  });
});