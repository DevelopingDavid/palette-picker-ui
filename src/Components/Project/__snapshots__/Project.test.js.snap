// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project should match Snapshot 1`] = `
<div>
  <WithStyles(ListItem)
    button={true}
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <i
        className="material-icons"
      >
        palette
      </i>
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      inset={true}
      primary="Project 1"
    />
    <WithStyles(ListItemIcon)>
      <i
        className="material-icons"
        onClick={[Function]}
      >
        delete
      </i>
    </WithStyles(ListItemIcon)>
    <pure(ExpandMoreIcon) />
  </WithStyles(ListItem)>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  />
  >
    <WithStyles(Typography)
      id="empty-project"
      variant="subtitle1"
    >
      This project is empty
    </WithStyles(Typography)>
  </WithStyles(Collapse)>
  <WithStyles(Divider) />
</div>
`;
